spikeinterface/core/addtemplatesrecording.py

```diff
@@ -0,0 +1,129 @@
+
+        if t_max is None:
+            if parent_recording is None:
+                t_max = [self.spike_vector['sample_ind'][-1] + templates.shape[1]]
+            else:
+                t_max = [parent_recording.get_num_frames(segment_index) for segment_index in range(sorting.get_num_segments())]
+
+
+        for segment_index in range(sorting.get_num_segments()):
+            start = np.searchsorted(self.spike_vector['segment_ind'], segment_index, side="left")
+            end = np.searchsorted(self.spike_vector['segment_ind'], segment_index, side="right")
+            spikes = self.spike_vector[start : end]
+
+            parent_recording_segment = None if parent_recording is None else parent_recording._recording_segments[segment_index]
+            recording_segment = AddTemplatesRecordingSegment(self.sampling_frequency, 0, self.dtype, spikes, templates, nbefore,
+                                                             amplitude_factor, parent_recording_segment, t_max[segment_index]) # TODO: t_start should not necessarily be 0.
+            self.add_recording_segment(recording_segment)
+
+
+
+class AddTemplatesRecordingSegment(BaseRecordingSegment):
+
+    def __init__(self, sampling_frequency: float, t_start: int, dtype, spike_vector: np.ndarray, templates: np.ndarray, nbefore: List[int],
+                 amplitude_factor: List[List[float]], parent_recording_segment: Union[BaseRecordingSegment, None] = None, t_max: Union[int, None] = None) -> None:
+
+        BaseRecordingSegment.__init__(self, sampling_frequency, t_start=t_start)
+        assert not (parent_recording_segment is None and t_max is None)
+
+        self.dtype = dtype
+        self.spike_vector = spike_vector
+        self.templates = templates
+        self.nbefore = nbefore
+        self.amplitude_factor = amplitude_factor
+        self.parent_recording = parent_recording_segment
+        self.t_max = parent_recording_segment.get_num_frames() if t_max is None else t_max
+
+    def get_traces(self, start_frame: Union[int, None] = None, end_frame: Union[int, None] = None,
+                   channel_indices: Union[List, None] = None) -> np.ndarray:
+        n_channels = self.templates.shape[2]
+        start_frame = 0 if start_frame is None else start_frame
+        end_frame = self.t_max if end_frame is None else end_frame
+        channel_indices = list(range(n_channels)) if channel_indices is None else channel_indices
+
+        if self.parent_recording is not None:
+            traces = self.parent_recording.get_traces(start_frame, end_frame, channel_indices)
+        else:
+            traces = np.zeros([end_frame - start_frame, n_channels], dtype=self.dtype)
+        
+
+        start = np.searchsorted(self.spike_vector['sample_ind'], start_frame - self.templates.shape[1], side="left")
+        end   = np.searchsorted(self.spike_vector['sample_ind'], end_frame   + self.templates.shape[1], side="right")
+        spike_vector = self.spike_vector[start : end]
+
```

Comment:
instead of the `spike_nb` business, can't you compute the amplitude factor for each spike here? I think it would be very quick with respect to the `for spike in spike_vector:` loop (and it would avoid modifications to `basesorting`!